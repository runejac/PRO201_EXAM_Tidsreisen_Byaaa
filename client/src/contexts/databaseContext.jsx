import React, { createContext } from "react";
import { fetchJSON_client } from "../helpers/http.jsx";

export const DatabaseContext = createContext({
  async listHistory() {
    return await fetchJSON_client("/api/history");
  },
  async listQuiz(query) {
    return await fetchJSON_client("/api/quiz?" + new URLSearchParams(query));
  },
  async listAudio() {
    return await fetchJSON_client("/api/sound");
  },
});
