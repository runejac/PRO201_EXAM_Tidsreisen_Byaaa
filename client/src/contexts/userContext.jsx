import React, { createContext } from "react";
import { fetchJSON_client, putJSON } from "../helpers/http.jsx";

export const UserContext = createContext({
  async getUser() {
    return await fetchJSON_client("/api/login");
  },
  async updateUser(user) {
    return await putJSON("/api/login/updateuser", user);
  },
});
