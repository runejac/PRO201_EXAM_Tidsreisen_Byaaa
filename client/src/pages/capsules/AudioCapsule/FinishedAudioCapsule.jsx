import React, { useContext, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import olafInfront from "../../../assets/images/olaf-infront.png";
import "../../../css/olaf.css";
import { CapsuleButtonGreen } from "../../../components/CapsuleButton.jsx";
import Notes from "./Notes";

const FinishedAudioCapsule = ({ update, filteredCapsule }) => {
  const navigate = useNavigate();

  // Updates database when hitting this component
  useEffect(async () => {
    await update();
  }, []);

  const navigateToMap = () => navigate("/map");
  const navigateToMyFindings = () => {
    navigate("/myfindings");
  };

  return (
    <div className="d-flex justify-content-center align-items-center flex-column vh-100 bg-capsule">
      <Notes />
      <div className={"d-flex justify-content-center mb-5"}>
        <h1 id={"done-title"}>Fullført lydkapselen!</h1>
      </div>
      <div
        id={"olaf-infront-div"}
        className={
          "d-flex align-content-center flex-row justify-content-center gap-3  position-relative z-index-1"
        }
      >
        <div>
          <img
            id={"olof-infront"}
            height={200}
            /*className={"img-fluid"}*/ src={olafInfront}
            alt="bilde av olaf på fullført side"
          />
        </div>

        <div className="pe-4">
          <div className="left-point"></div>
          {filteredCapsule ? (
            <p id={"finish-paragraph"}>
              Woops! Det ser ut som du allerede har vært på denne kapselen. Gå
              til neste kapsel.
            </p>
          ) : (
            <p id={"finish-paragraph"}>
              Godt jobbet!
              <br /> Du har gjennomført en kapsel og fått poeng for det, og
              kanskje til og med lært noe nytt om Byåa!
            </p>
          )}
        </div>
      </div>
      {filteredCapsule ? (
        <div>
          <p style={{ fontSize: "1.7rem", fontWeight: "bold" }}></p>
        </div>
      ) : (
        <div>
          <p style={{ fontSize: "1.7rem", fontWeight: "bold" }}>
            Du har fått + 20 poeng
          </p>
        </div>
      )}

      <div className={"d-flex flex-column"}>
        <div>
          <CapsuleButtonGreen
            onClick={navigateToMap}
            buttonText={"Tilbake til kart"}
          />
        </div>
        <div>
          <CapsuleButtonGreen
            onClick={navigateToMyFindings}
            buttonText={"Mine funn"}
          />
        </div>
      </div>
    </div>
  );
};

export default FinishedAudioCapsule;
