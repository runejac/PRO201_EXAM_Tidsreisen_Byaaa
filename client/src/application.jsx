import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
