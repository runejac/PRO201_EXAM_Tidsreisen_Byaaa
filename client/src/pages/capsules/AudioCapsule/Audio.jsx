import React, { useState, useContext, useEffect } from "react";
import Play from "./Play";
import FinishedAudioCapsule from "./FinishedAudioCapsule.jsx";
import { useLoading } from "../../../helpers/useLoading";
import { DatabaseContext } from "../../../contexts/databaseContext";
import { useParams } from "react-router-dom";
import { User } from "../../../application.jsx";
import { NotLoggedIn } from "../../../components/NotLoggedIn.jsx";
import { UserContext } from "../../../contexts/userContext.jsx";
import { AiFillSound } from "react-icons/ai";
import { Loading } from "../../../components/Loading";
import "../../../css/sound.css";
import Notes from "./Notes";

const Audio = () => {
  const { id } = useParams();
  const [drag, setDrag] = useState();
  const [count, setCount] = useState(0);
  const { updateUser } = useContext(UserContext);
  const { user, setUser } = useContext(User);
  const { name } = user;
  const { listAudio } = useContext(DatabaseContext);
  const { data, loading } = useLoading(async () => await listAudio());
  const [songInfo, setSongInfo] = useState({
    currentTime: 0,
    duration: null,
  });

  // updates correct user only, and only if count is more than 0
  useEffect(() => {
    if (count > 0) {
      return () => {
        if (!user.finishedCapsules.includes(capsuleObject)) {
          user.finishedCapsules.push(capsuleObject);
          user.points = user.points + 20;

          setUser({
            ...user,
          });
        }
      };
    }
  }, [count]);

  // updates user with capsuleObject and points
  const updateToDatabase = async () => {
    setCount(count + 1);
    await updateUser({
      user,
      finishedCapsules: capsuleObject,
      points: 20,
    });
  };
  // If the site is loading, we're setting the loading screen
  if (loading) {
    return <Loading />;
  }

  // If name is undefiend we're showing the not logged in screen
  if (name === undefined) {
    return <NotLoggedIn />;
  }

  // Filtering out right data for the capsuleObject
  const getId = data
    ?.map((id) => id)
    .filter((item) => item.title.toLowerCase() === id?.toLowerCase())
    .map((itemId) => itemId.id);

  // setting capsule object to be inserted in database
  const capsuleObject = {
    id: getId.toString(),
    name: id,
    category: "Lydkapsel",
  };

  // checks capsule id from database
  const capsuleNameFromDatabase = user.finishedCapsules.map((capsuleName) => {
    return capsuleName.name;
  });

  // finds capsule id matches to check whether if user has done it or not
  const filteredCapsuleNamesFromUserDatabase = capsuleNameFromDatabase.find(
    (capsuleName) => {
      console.log(capsuleName);
      return capsuleName === id;
    }
  );
  return (
    <>
      {songInfo.duration === songInfo.currentTime ? (
        <FinishedAudioCapsule
          filteredCapsule={filteredCapsuleNamesFromUserDatabase}
          update={updateToDatabase}
        />
      ) : (
        <div className="position-relative d-flex justify-content-center align-items-center flex-column vh-100 bg-capsule media">
          <Notes />
          {data?.map((item, index) => {
            return (
              <div key={index}>
                {item.title.toLowerCase() === id?.toLowerCase() && (
                  <div className="position-relative d-flex justify-content-center align-items-center flex-column ">
                    <h3 className="pb-3 d-flex align-items-center title">
                      <AiFillSound color={"var(--textColorGray)"} />{" "}
                      {item.category}
                    </h3>
                    <h1 className="pb-3 fw-bolder text-capsule subtitle">
                      {item.title}
                    </h1>
                    <img
                      className="p-2 img"
                      src={item.image}
                      alt={item.image}
                    />

                    <Play
                      song={item.song}
                      songInfo={songInfo}
                      setSongInfo={setSongInfo}
                      setDrag={setDrag}
                    />
                    <h4 className="my-5 fw-bold year">År {item.year}</h4>
                    {filteredCapsuleNamesFromUserDatabase ? (
                      <p className="fst-italic fw-bold undertext">
                        Denne lydkapselen er lyttet ferdig
                      </p>
                    ) : (
                      <p className="fst-italic fw-bold undertext">
                        Lytt ferdig lydkapselen for å få poeng
                      </p>
                    )}
                  </div>
                )}
              </div>
            );
          })}
        </div>
      )}
    </>
  );
};

export default Audio;
