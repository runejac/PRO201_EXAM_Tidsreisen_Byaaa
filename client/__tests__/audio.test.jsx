/**
 * @jest-environment jsdom
 */

import { act } from "react-dom/test-utils";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import { DatabaseContext } from "../src/contexts/databaseContext.jsx";
import { User } from "../src/application.jsx";
import { UserContext } from "../src/contexts/userContext.jsx";
import Audio from "../src/pages/capsules/AudioCapsule/Audio.jsx";
import FinishedAudioCapsule from "../src/pages/capsules/AudioCapsule/FinishedAudioCapsule.jsx";

describe("Audio test", function () {
  it("should render audio component", async function () {
    const user = {
      name: "Hei",
      points: 23,
      finishedCapsules: [],
      intro: true,
      walk: false,
    };

    const audio = [
      {
        category: "Lydkapsel",
        title: "test",
        year: "2000",
        image: "image",
        id: "1",
        song: "songtitle",
      },
    ];

    const context = jest.fn();
    const container = document.createElement("div");

    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <DatabaseContext.Provider value={{ listAudio: () => audio }}>
            <User.Provider value={{ user }}>
              <UserContext.Provider value={{ context }}>
                <Audio />
              </UserContext.Provider>
            </User.Provider>
          </DatabaseContext.Provider>
        </MemoryRouter>,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });

  it("should render FinishedAudioCapsule", function () {
    const container = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <FinishedAudioCapsule />
      </MemoryRouter>,
      container
    );
    expect(container).toMatchSnapshot();
  });
});
