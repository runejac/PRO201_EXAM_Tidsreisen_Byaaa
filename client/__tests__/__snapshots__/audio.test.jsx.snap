// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio test should render FinishedAudioCapsule 1`] = `
<div>
  <div
    class="d-flex justify-content-center align-items-center flex-column vh-100 bg-capsule"
  >
    <div>
      <div
        style="position: absolute; top: 150px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
        />
      </div>
      <div
        style="position: absolute; bottom: 20px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="100"
        />
      </div>
      <div
        style="position: absolute; top: 200px; right: 20px; color: rgba(255, 255, 255, 0.271); transform: scaleX(-1);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="80"
        />
      </div>
      <div
        style="position: absolute; top: 500px; right: 20px; color: rgba(255, 255, 255, 0.271); transform: scaleX(-1);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="50"
        />
      </div>
      <div
        style="position: absolute; top: 400px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="50"
        />
      </div>
    </div>
    <div
      class="d-flex justify-content-center mb-5"
    >
      <h1
        id="done-title"
      >
        Fullført lydkapselen!
      </h1>
    </div>
    <div
      class="d-flex align-content-center flex-row justify-content-center gap-3  position-relative z-index-1"
      id="olaf-infront-div"
    >
      <div>
        <img
          alt="bilde av olaf på fullført side"
          height="200"
          id="olof-infront"
          src="test-file-stub"
        />
      </div>
      <div
        class="pe-4"
      >
        <div
          class="left-point"
        />
        <p
          id="finish-paragraph"
        >
          Godt jobbet!
          <br />
           Du har gjennomført en kapsel og fått poeng for det, og kanskje til og med lært noe nytt om Byåa!
        </p>
      </div>
    </div>
    <div>
      <p
        style="font-size: 1.7rem; font-weight: bold;"
      >
        Du har fått + 20 poeng
      </p>
    </div>
    <div
      class="d-flex flex-column"
    >
      <div>
        <button
          class="my-2"
          id="button-capsules-green"
        >
          Tilbake til kart
        </button>
      </div>
      <div>
        <button
          class="my-2"
          id="button-capsules-green"
        >
          Mine funn
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Audio test should render audio component 1`] = `
<div>
  <div
    class="position-relative d-flex justify-content-center align-items-center flex-column vh-100 bg-capsule media"
  >
    <div>
      <div
        style="position: absolute; top: 150px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
        />
      </div>
      <div
        style="position: absolute; bottom: 20px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="100"
        />
      </div>
      <div
        style="position: absolute; top: 200px; right: 20px; color: rgba(255, 255, 255, 0.271); transform: scaleX(-1);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="80"
        />
      </div>
      <div
        style="position: absolute; top: 500px; right: 20px; color: rgba(255, 255, 255, 0.271); transform: scaleX(-1);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="50"
        />
      </div>
      <div
        style="position: absolute; top: 400px; left: 20px; color: rgba(255, 255, 255, 0.271);"
      >
        <img
          alt="Some note"
          src="test-file-stub"
          width="50"
        />
      </div>
    </div>
    <div />
  </div>
</div>
`;
